Data/Aeson/Parser/Internal.hs

```diff
@@ -98,23 +120,38 @@ json = value
 -- This function is an alias for 'value''. In aeson 0.8 and earlier, it
 -- parsed only object or array types, in conformance with the
 -- now-obsolete RFC 4627.
+--
+-- ==== Warning
+--
+-- If an object contains duplicate keys, only the first one will be kept.
+-- For a more flexible alternative, see 'jsonWith''.
 json' :: Parser Value
 json' = value'
 
-object_ :: Parser Value
-object_ = {-# SCC "object_" #-} Object <$> objectValues jstring value
+-- Open recursion: object_, object_', array_, array_' are parameterized by the
+-- toplevel Value parser to be called recursively, to keep the parameter
+-- hFromList outside of the recursive loop for proper inlining.
+
+object_ :: HFromList -> Parser Value -> Parser Value
+object_ hFromList val = {-# SCC "object_" #-} Object <$> objectValues hFromList jstring val
+{-# INLINE object_ #-}
 
-object_' :: Parser Value
-object_' = {-# SCC "object_'" #-} do
-  !vals <- objectValues jstring' value'
+object_' :: HFromList -> Parser Value -> Parser Value
+object_' hFromList val' = {-# SCC "object_'" #-} do
+  !vals <- objectValues hFromList jstring' val'
   return (Object vals)
  where
   jstring' = do
     !s <- jstring
     return s
+{-# INLINE object_' #-}
 
-objectValues :: Parser Text -> Parser Value -> Parser (H.HashMap Text Value)
-objectValues str val = do
+-- The object parser 'objectValues' is parameterized by the object constructor.
+-- See also 'jsonWith'.
+type HFromList = [(Text, Value)] -> Parser Object
```

Comment:
I'd have `Either String Object` as a return type. `Parser` is "too big".

And actually wouldn't use a type synonym, `hFromList` is confusing naming, `mkObject` is what it is

Reply:
I see. I forgot we're in attoparsec's `Parser`, not aeson's `Parser` (which is closer to `Either`). That's a fair point about the synonym, I'll remove it. Thanks for the review!